File: src/utils/serviceOrderPricing.js
/**
 * Service-order pricing CMS (mirrors backend/service_order_pricing.py).
 * Tokens for copy: {{trap}}, {{from}}, {{p50_21}}, {{p100_7}}, …
 */

export const DEFAULT_SERVICE_ORDER_PRICING = {
  deadlines: [
    { days: 21, label: '2–3 недели', hint: '14–21 день', price_50: 25000, price_100: 20000 },
    { days: 10, label: '1–2 недели', hint: '8–13 дней', price_50: 30000, price_100: 25000 },
    { days: 7, label: 'Неделя', hint: '7 дней', price_50: 35000, price_100: 30000 },
    { days: 3, label: '2–3 дня', hint: 'быстрее', price_50: 40000, price_100: 35000 },
    { days: 1, label: '24 часа', hint: 'срочно', price_50: 50000, price_100: 45000 },
  ],
  trap_price: 15000,
  copy: {
    guide_title: 'Прайс услуг',
    guide_subtitle: 'от {{from}} · срок и предоплата меняют цену',
    song_title: 'Песня под ключ',
    song_body:
      'Песня с мелодиями и текстом (текст опционально). Права — заказчику, без указания авторства.',
    trap_title: 'Бит в стиле трэп',
    trap_body: 'Фикс. цена {{trap}}, срок на стоимость не влияет.',
    col_50_title: 'Предоплата 50%',
    col_100_title: 'Оплата 100%',
  },
};

export function formatServiceRub(amount) {
  const n = Number(amount);
  if (!Number.isFinite(n)) return '0 ₽';
  // Regular spaces — easier for templates / copy-paste than locale NBSP
  return `${Math.round(n).toString().replace(/\B(?=(\d{3})+(?!\d))/g, ' ')} ₽`;
}

function asInt(value, fallback) {
  const n = Number(value);
  if (!Number.isFinite(n)) return fallback;
  return Math.max(0, Math.round(n));
}

export function normalizeServiceOrderPricing(raw) {
  const base = structuredClone
    ? structuredClone(DEFAULT_SERVICE_ORDER_PRICING)
    : JSON.parse(JSON.stringify(DEFAULT_SERVICE_ORDER_PRICING));

  if (!raw || typeof raw !== 'object') return base;

  base.trap_price = asInt(raw.trap_price, base.trap_price);

  if (Array.isArray(raw.deadlines) && raw.deadlines.length) {
    const rows = [];
    for (let i = 0; i < raw.deadlines.length; i += 1) {
      const item = raw.deadlines[i];
      if (!item || typeof item !== 'object') continue;
      const fallback = base.deadlines[Math.min(i, base.deadlines.length - 1)];
      let days = asInt(item.days, fallback.days);
      if (days <= 0) days = fallback.days;
      rows.push({
        days,
        label: String(item.label || fallback.label).trim().slice(0, 80) || fallback.label,
        hint: String(item.hint ?? fallback.hint).trim().slice(0, 80),
        price_50: asInt(item.price_50, fallback.price_50),
        price_100: asInt(item.price_100, fallback.price_100),
      });
    }
    if (rows.length) base.deadlines = rows.slice(0, 12);
  }

  if (raw.copy && typeof raw.copy === 'object') {
    Object.keys(base.copy).forEach((key) => {
      if (raw.copy[key] != null) base.copy[key] = String(raw.copy[key]).slice(0, 2000);
    });
  }

  return base;
}

/** Chips for non-dev admins: label → token, grouped for CMS UI */
export function listPriceVariableChips(pricing) {
  const data = normalizeServiceOrderPricing(pricing);
  const chips = [
    { token: '{{trap}}', label: 'Трэп-бит', group: 'base', groupLabel: 'Базовые' },
    { token: '{{from}}', label: 'Минимум (от …)', group: 'base', groupLabel: 'Базовые' },
  ];
  data.deadlines.forEach((row) => {
    chips.push({
      token: `{{p50_${row.days}}}`,
      label: row.label,
      group: 'p50',
      groupLabel: 'Предоплата 50%',
    });
    chips.push({
      token: `{{p100_${row.days}}}`,
      label: row.label,
      group: 'p100',
      groupLabel: 'Оплата 100%',
    });
  });
  return chips;
}

/** Group chips preserving order of first appearance. */
export function groupPriceVariableChips(chips) {
  const list = Array.isArray(chips) ? chips : [];
  const order = [];
  const map = new Map();
  for (const chip of list) {
    const key = chip.group || 'other';
    if (!map.has(key)) {
      map.set(key, { group: key, groupLabel: chip.groupLabel || key, chips: [] });
      order.push(key);
    }
    map.get(key).chips.push(chip);
  }
  return order.map((key) => map.get(key));
}

export function buildPriceVars(pricing) {
  const data = normalizeServiceOrderPricing(pricing);
  const vars = { trap: formatServiceRub(data.trap_price) };
  const amounts = [data.trap_price];
  data.deadlines.forEach((row) => {
    vars[`p50_${row.days}`] = formatServiceRub(row.price_50);
    vars[`p100_${row.days}`] = formatServiceRub(row.price_100);
    amounts.push(row.price_50, row.price_100);
  });
  vars.from = formatServiceRub(Math.min(...amounts));
  return vars;
}

export function substitutePriceVars(text, pricing) {
  if (!text) return '';
  const vars = buildPriceVars(pricing);
  return String(text).replace(/\{\{([a-zA-Z0-9_]+)\}\}/g, (full, key) =>
    Object.prototype.hasOwnProperty.call(vars, key) ? vars[key] : full
  );
}

export function getServicePrice(pricing, deadlineDays, prepaymentPercent) {
  const data = normalizeServiceOrderPricing(pricing);
  const days = parseInt(deadlineDays, 10);
  if (!Number.isFinite(days)) return null;
  const pct = Number(prepaymentPercent) >= 100 ? 100 : 50;
  const key = pct === 100 ? 'price_100' : 'price_50';

  const exact = data.deadlines.find((r) => r.days === days);
  if (exact) return exact[key];

  const rows = [...data.deadlines].sort((a, b) => a.days - b.days);
  if (!rows.length) return null;
  if (days >= 14) {
    const hit = rows.find((r) => r.days >= 14) || rows[rows.length - 1];
    return hit[key];
  }
  if (days === 1) {
    const hit = rows.find((r) => r.days === 1) || rows[0];
    return hit[key];
  }
  return rows.reduce((best, row) =>
    Math.abs(row.days - days) < Math.abs(best.days - days) ? row : best
  )[key];
}

export function resolvedServiceCopy(pricing) {
  const data = normalizeServiceOrderPricing(pricing);
  const out = {};
  Object.entries(data.copy).forEach(([k, v]) => {
    out[k] = substitutePriceVars(v, data);
  });
  return out;
}

File: backend/service_order_pricing.py
"""Service-order pricing CMS (site_settings.service_order_pricing)."""
from __future__ import annotations

import json
import re
from copy import deepcopy
from typing import Any

DEFAULT_SERVICE_ORDER_PRICING: dict[str, Any] = {
    "deadlines": [
        {
            "days": 21,
            "label": "2–3 недели",
            "hint": "14–21 день",
            "price_50": 25000,
            "price_100": 20000,
        },
        {
            "days": 10,
            "label": "1–2 недели",
            "hint": "8–13 дней",
            "price_50": 30000,
            "price_100": 25000,
        },
        {
            "days": 7,
            "label": "Неделя",
            "hint": "7 дней",
            "price_50": 35000,
            "price_100": 30000,
        },
        {
            "days": 3,
            "label": "2–3 дня",
            "hint": "быстрее",
            "price_50": 40000,
            "price_100": 35000,
        },
        {
            "days": 1,
            "label": "24 часа",
            "hint": "срочно",
            "price_50": 50000,
            "price_100": 45000,
        },
    ],
    "trap_price": 15000,
    "copy": {
        "guide_title": "Прайс услуг",
        "guide_subtitle": "от {{from}} · срок и предоплата меняют цену",
        "song_title": "Песня под ключ",
        "song_body": (
            "Песня с мелодиями и текстом (текст опционально). "
            "Права — заказчику, без указания авторства."
        ),
        "trap_title": "Бит в стиле трэп",
        "trap_body": "Фикс. цена {{trap}}, срок на стоимость не влияет.",
        "col_50_title": "Предоплата 50%",
        "col_100_title": "Оплата 100%",
    },
}

_TOKEN_RE = re.compile(r"\{\{([a-zA-Z0-9_]+)\}\}")


def _as_int(value: Any, default: int) -> int:
    try:
        n = int(float(value))
    except (TypeError, ValueError):
        return default
    return max(0, n)


def format_rub(amount: int) -> str:
    return f"{amount:,}".replace(",", " ") + " ₽"


def build_price_vars(pricing: dict) -> dict[str, str]:
    """Map template tokens → formatted rubles for admin chips / public copy."""
    data = normalize_service_order_pricing(pricing)
    vars_map: dict[str, str] = {
        "trap": format_rub(data["trap_price"]),
    }
    amounts = [data["trap_price"]]
    for row in data["deadlines"]:
        days = row["days"]
        p50 = row["price_50"]
        p100 = row["price_100"]
        vars_map[f"p50_{days}"] = format_rub(p50)
        vars_map[f"p100_{days}"] = format_rub(p100)
        amounts.extend([p50, p100])
    vars_map["from"] = format_rub(min(amounts) if amounts else 0)
    return vars_map


def substitute_price_vars(text: str, pricing: dict) -> str:
    if not text:
        return ""
    vars_map = build_price_vars(pricing)

    def repl(match: re.Match) -> str:
        key = match.group(1)
        return vars_map.get(key, match.group(0))

    return _TOKEN_RE.sub(repl, str(text))


def normalize_service_order_pricing(raw: Any) -> dict[str, Any]:
    base = deepcopy(DEFAULT_SERVICE_ORDER_PRICING)
    if not isinstance(raw, dict):
        return base

    trap = _as_int(raw.get("trap_price"), base["trap_price"])
    base["trap_price"] = trap

    raw_deadlines = raw.get("deadlines")
    if isinstance(raw_deadlines, list) and raw_deadlines:
        normalized_rows = []
        for i, item in enumerate(raw_deadlines):
            if not isinstance(item, dict):
                continue
            fallback = base["deadlines"][min(i, len(base["deadlines"]) - 1)]
            days = _as_int(item.get("days"), fallback["days"])
            if days <= 0:
                days = fallback["days"]
            label = str(item.get("label") or fallback["label"]).strip() or fallback["label"]
            hint = str(item.get("hint") if item.get("hint") is not None else fallback["hint"]).strip()
            normalized_rows.append(
                {
                    "days": days,
                    "label": label[:80],
                    "hint": hint[:80],
                    "price_50": _as_int(item.get("price_50"), fallback["price_50"]),
                    "price_100": _as_int(item.get("price_100"), fallback["price_100"]),
                }
            )
        if normalized_rows:
            base["deadlines"] = normalized_rows[:12]

    raw_copy = raw.get("copy")
    if isinstance(raw_copy, dict):
        for key in base["copy"].keys():
            if key in raw_copy and raw_copy[key] is not None:
                base["copy"][key] = str(raw_copy[key])[:2000]

    return base


def parse_service_order_pricing_json(raw: str) -> dict[str, Any]:
    if not (raw or "").strip():
        return deepcopy(DEFAULT_SERVICE_ORDER_PRICING)
    try:
        data = json.loads(raw)
    except json.JSONDecodeError:
        return deepcopy(DEFAULT_SERVICE_ORDER_PRICING)
    return normalize_service_order_pricing(data)


def get_price_for_deadline(pricing: dict, deadline_days: int, prepayment_percent: int) -> int | None:
    data = normalize_service_order_pricing(pricing)
    days = int(deadline_days)
    pct = 100 if int(prepayment_percent) >= 100 else 50
    key = "price_100" if pct == 100 else "price_50"

    exact = next((r for r in data["deadlines"] if r["days"] == days), None)
    if exact:
        return exact[key]

    # Fallback ranges matching legacy OrderPage logic
    rows = sorted(data["deadlines"], key=lambda r: r["days"])
    if not rows:
        return None
    if days >= 14:
        hit = next((r for r in rows if r["days"] >= 14), rows[-1])
        return hit[key]
    if days == 1:
        hit = next((r for r in rows if r["days"] == 1), rows[0])
        return hit[key]
    # nearest by days
    hit = min(rows, key=lambda r: abs(r["days"] - days))
    return hit[key]
